### PR TITLE
Fix for showon !: handling

### PR DESCRIPTION
**Pull request by GeraintEdwards**
Resolves issue https://github.com/joomla/joomla-cms/issues/14257
Corrects javascript coding errors in https://github.com/joomla/joomla-cms/pull/13476

Pull Request for Issue #14257 .

### Summary of Changes

There are logical errors in the javascript in linkedoptions in cms-uncompressed.js that stop the !: showon operator from working in most cases.

1. The primary problem is that `for (var i in itemval)` loops over non-enumerable properties and prototypes which means that != will ALWAYS match 
2. secondary issue is that multiple select tags get a null from `itemval = $field.val();` which causes problems.
3. radio and checkboxes return a blank value in the current code which always matches a != condition

### Testing Instructions

Create a form with this XML 
```javascript
<field name="masterfield" type="radio" default="1" class="btn-group btn-group-yesno" label="MasterField" >
	<option value="0">unknown</option>
	<option value="1">value 1</option>
	<option value="2">value 2</option>
</field>
<field name="subfield" type="radio" default="1" showon="masterfield!:0" class="btn-group btn-group-yesno" label="SubField" >
	<option value="1">Show text field</option>
	<option value="2">Show URL field</option>
</field>
<field name="subText" showon="masterfield!:0[AND]subfield:1" type="text" default="" label="Text - should appear then MasterField is !unknown and subfield is text" />
<field name="subURL" showon="masterfield!:0[AND]subfield:2" size="20" type="url" default="" label="URL - should appear then MasterField is !unknown and subfield is url" />        

<field name="selectlist" type="list" multiple="multiple" default="0" label="Multiple Select Conditions" >
	<option value="0">select a value</option>
	<option value="1">value 1</option>
	<option value="2">value 2</option>
	<option value="3">value 3</option>
	<option value="4">value 4</option>
</field>
<field name="extratext" showon="selectlist!:0" type="text" default="" label="Text should appear when something other than 'select a value' is selected" />

<field name="selectlist2" type="list" default="0" label="Single Select Conditions" >
	<option value="0">select a value</option>
	<option value="1">value 1</option>
	<option value="2">value 2</option>
	<option value="3">value 3</option>
	<option value="4">value 4</option>
</field>
<field name="extratext2" showon="selectlist2!:0" type="text" default="" label="Text should appear when something other than 'select a value' is selected" />

```

For Masterfield - choose 1 or 2 
For SubField choose text or url

Multiple Select Conditions - choose
1. no values, 
2. 'select a value' and "value 1"
3, just 'select a value'
4. just "value 1"

Single Select Conditions - choose
1.  'select a value'
2. "value 1"

### Expected result

The Text or URL fields should appear and switch as you select SubField values

Multiple Select Conditions:
1. 'text' field should be visible
2. 'text' field should be visible
3. 'text' field should NOT be visible
4. 'text' field should be visible

Single Select Conditions
1. 'text' field should NOT be visible
2. 'text' field should be visible

### Actual result

The  text and url fields  are not affected by the value of Masterfield

Multiple Select Conditions:
1. 'text' field is visible - correct
2. 'text' field is visible - correct
3. 'text' field is visible -  INCORRECT
4. 'text' field is visible - correct

Single Select Conditions
1. 'text' field is visible -  INCORRECT
2. 'text' field is visible -  correct

### Documentation Changes Required

p.s. cms-compressed.js will also need changing but I don't know how to generate this.

### CODE REVIEW

**Comment by brianteeman** [comment]
Can you also add the compressed version of the file please.

**Comment by GeraintEdwards** [comment]
Brian - I'm afraid I don't know how that is done.  I hunted high and low in Google to no avail.  Do you know what the official tool it?

In the meantime the fix can be tested in debug mode.

**Comment by brianteeman** [comment]
there is no official tool - just use any minifier

**Comment by GeraintEdwards** [comment]
Added updated cms.js (wasn't the same minify tool as was used last time though).

**Comment by tonypartridge** [comment]
@franz-wohlkoenig would you have time to test this?

**Comment by franz-wohlkoenig** [comment]
@tonypartridge don't understand Instructions like "Create a form with this XML".

**Comment by tonypartridge** [comment]
@franz-wohlkoenig so it means A basic component which uses JForm use the xml to generate the form.

**Comment by franz-wohlkoenig** [comment]
there lacks my Knowledge too. I can test having clear instructions like "click on this and look if that happens, "A basic component which uses JForm use the xml to generate the form" is unclear.

**Comment by GeraintEdwards** [comment]
You can add the XML to any Joomla component configuration file e.g. add it to the start of the first fieldset of administrator/components/com_contact/config.xml and then open the com_contact options/config page - you'll see the new fields at the top.

File: media/jui/js/cms-uncompressed.js [hunk 1]
```diff
@@ -91,9 +101,32 @@ if (!Array.prototype.indexOf)
 						itemval = JSON.parse('["' + itemval + '"]');
 					}
 
+                                        // for (var i in itemval) loops over non-enumerable properties and prototypes which means that != will ALWAYS match 
+                                        // see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in
+                                        // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyNames
+                                        // use native javascript Array forEach - see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach                                        
+                                        itemval.forEach(function (itemval_val, i)
```
Comment (C-Lodder) [inline media/jui/js/cms-uncompressed.js]: Can't use `forEach` as this isn't supported by IE8

Comment (GeraintEdwards) [inline media/jui/js/cms-uncompressed.js]: Is my alternative code ok (commented out at line 124 onwards)?  Its not quite as neat but works in the same way - the critical thing is to only loop over the enumerable elements.

Comment (wilsonge) [inline media/jui/js/cms-uncompressed.js]: @C-Lodder bump on this

**Comment by C-Lodder** [comment]
This is working fine based on the test you've given.

As it stands, the `OR` clause works fine if you're targetting the same field, like so:

```
<field name="field1" type="radio" default="1" class="btn-group btn-group-yesno" label="field1">
	<option value="0">0</option>
	<option value="1">1</option>
</field>
<field name="field2" type="radio" default="1" class="btn-group btn-group-yesno" label="field2">
	<option value="0">0</option>
	<option value="1">1</option>
	<option value="2">2</option>
</field>
<field name="field3" showon="field1:1[AND]field2:1[OR]field2:2" type="text" default="" label="field3" />
``` 

(If `field2` is equal to `1` OR `2`)

But, let's say I have 4 fields:

```
<field name="field1" type="radio" default="1" class="btn-group btn-group-yesno" label="field1">
	<option value="0">0</option>
	<option value="1">1</option>
</field>
<field name="field2" type="radio" default="1" class="btn-group btn-group-yesno" label="field2">
	<option value="0">0</option>
	<option value="1">1</option>
	<option value="2">2</option>
</field>
<field name="field3" type="radio" default="1" class="btn-group btn-group-yesno" label="field3">
	<option value="0">0</option>
	<option value="1">1</option>
</field>
<field name="field4" showon="field1:1[AND]field2:1[OR]field3:1" type="text" default="" label="field4" />
```

Basically, I only want to show `field4` if:

- `field1` = `1` AND
- `field2` = `1` OR
- `field3` = `1`

Not sure if we want to go this deep or not, but just another case to take into consideration.

**Comment by GeraintEdwards** [comment]
The problem with [AND] with [OR] is that what is being tested is ambiguous.

Does mean `{field1:1[AND]field2:1} [OR]field3:1` or does it mean `field1:1[AND] {field2:1[OR]field3:1}` or `{field1:1[AND]field2:1} [OR]field3:1}`

I don't think we would want to handle parenthesis (???) but we could make a documented assumption about the sequence of testing or the priority of operators.

So you decide to work from left to right then  `field1:1[AND]field2:1[OR]field3:1` would mean `{field1:1[AND]field2:1} [OR]field3:1}` and if you wanted `field1:1[AND] {field2:1[OR]field3:1}`  you would use `field2:1[OR]field3:1[AND]field1:1` in the XML file.

Or you could say that the [AND]s are all tested before the [OR]s (which is less flexible).

My main objective was to resolve the issues with the not (!) operator and handling values from multiple-choice select boxes.

After this PR is accepted then maybe a review of the functionality would be appropriate.  For example I've noticed that showons in subforms do not take account of values from the parent form.

**Comment by brianteeman** [comment]
and or is always complex and subject to opinion - i would suggest keeping this simple for now and stick to the original objectives

**Comment by C-Lodder** [comment]
I have tested this item :white_check_mark: successfully on d2d41eac64c6f23e4566a66b3ebc21144274a8f5<br /><br />Yup, that's perfectly fine :) In which case:<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/15104">issues.joomla.org/tracker/joomla-cms/15104</a>.</sub>

**Comment by tonypartridge** [comment]
I have tested this item :white_check_mark: successfully on d2d41eac64c6f23e4566a66b3ebc21144274a8f5<br /><br />Tested with the original scenario given. Works on current suggested usage case, agree if we want more complexity we should open an issue for discussion on the preferred method. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/15104">issues.joomla.org/tracker/joomla-cms/15104</a>.</sub>

**Comment by franz-wohlkoenig** [comment]
RTC after two successful tests.

**Comment by infograf768** [comment]
fyi, while making a pr i found out that showon was not working for filter.xml (searchtools).
for volunteers to look at 😏

**Comment by brianteeman** [comment]
I assume you meant here https://github.com/joomla/joomla-cms/issues/15621#issuecomment-297713146 or did you find something else

**Comment by infograf768** [comment]
was indeed for modules filters

**Comment by GeraintEdwards** [comment]
@infograf768 I'm happy to take a look at the showon/filter.xml issue but the issue referenced by Brian is closed.  Can you elaborate on the issue you have found?

**Comment by brianteeman** [comment]
@GeraintEdwards it is closed because JM wrote a different solution to work around that bug. 

If you look at 3.7.0 you can see https://github.com/joomla/joomla-cms/blob/3.7.0/administrator/components/com_modules/models/forms/filter_modules.xml#L55

and to put it simply that line did nothing

**Comment by GeraintEdwards** [comment]
Got you - thanks.  I just checked JM's changes and this showon fix would also have resolved the issue for the page filter display though it may not have helped any search values in the session/registry when switching from site to admin.

**Comment by infograf768** [comment]
What I did was to unset here
https://github.com/joomla/joomla-cms/pull/16118

**Pull request merged**